### DISCUSSION TITLE
getPixelAt

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#173] [Q&A]
Hi,
I have "borrowed" ;-) this code from the well respected member **dnaldoog**, which he posted September 25, 2020 at 7:53 am on the old Ctrlr forum. It works and draws an image on an **uiCustom Component**. I need to find the colour of pixels to change them if needed, and it seems logical to use getPixalAt(x,y) in the image. But (as usually) I don't seem to get it working -:(

 I know that  **a = Image:getPixelAt(30,30)** is wrong, but I don't know what is the right code. TIA!

function graphicsLCD(mod,g)
	frResource=resources:getResource("KurzScreen")
	if frResource~=nil then
		frImage	= Image()
		frImage = frResource:asImage()
        g:setOpacity(1.0)  -- 0.0 - 1.0		
        g:drawImageAt(frImage,0,0,false) -- x,y position in uiComponent
	end -- image found
	a = Image:getPixelAt(30,30)
	console("color :"..a)
end

**Comment by dnaldoog** [comment]
https://ctrlr.org/forums/topic/looking-to-change-button-image-on-value-change/#post-120030

Hi there @bijlevel,

I realised there's some redundancy in the code:

<pre>
frResource=resources:getResource("KurzScreen")
frImage = Image() -- gets overwritten by next line anyway!!
frImage = frResource:asImage()
</pre>

I have tested this using the attached panel which I created back in September (_now modified_):

<pre>
function graphicsLCD(mod, g)
    frImage = Image(resources:getResourceAsImage("KurzScreen"))
    if frImage ~= nil then
        g:setOpacity(1.0) -- 0.0 - 1.0
        g:drawImageAt(frImage, 0, 0, false) -- x,y position in uiComponent
        console(Colour(frImage:getPixelAt(30, 30)):toString()) -- will return hex string
    end -- image found
end

</pre>

[uiComponentAsButton_1_0_Hell-O-Kitty_2021-02-19_08-43.zip](https://github.com/RomanKubiak/ctrlr/files/6006676/uiComponentAsButton_1_0_Hell-O-Kitty_2021-02-19_08-43.zip)

See [https://github.com/RomanKubiak/ctrlr/](https://github.com/RomanKubiak/ctrlr/blob/7fab6623d3a2b26b1bcc583568028552936fd487/Source/Lua/JuceClasses/LGraphics.cpp)

**Reply by bijlevel** [reply]
Hi John,
Thank you so much! Just curious, how did you (and many of the people on this forum) acquire so much knowledge about programming Ctrlr? Can you recommend some reading? (I know that the old forum is a big source of knowledge, I do visit it frequently, but finding exact solutions for my programming issues is very time consuming.) For me the most difficult thing is to understand the relation between LUA and JUCE, primarily how it is related syntactically. To me the methods used in Ctrlr seem a combination of LUA syntax and JUCE syntax and it looks like sometimes one has to use JUCE syntax and sometimes LUA syntax, but it is hard to know when to use either of them or how to combine it. If this is made clear to me, the I can search more targeted instead of finding solutions accidentally or bothering this forum -;)

**Comment by dnaldoog** [comment]
Hi @bijlevel 

I often refer to the JUCE documentation to help me and the source code for Ctrlr. When you find a method called .def(...)  in Ctrlr source code it  means a JUCE C++ function/method is bound  to lua  (i.e can be used in lua) -  (using luabind). The next challenge is to find the correct syntax. That's when I use the [Monkey Theorum!](https://en.wikipedia.org/wiki/Infinite_monkey_theorem) LoL

When I found [.def("getPixelAt", &Image::getPixelAt)](https://github.com/RomanKubiak/ctrlr/blob/7fab6623d3a2b26b1bcc583568028552936fd487/Source/Lua/JuceClasses/LGraphics.cpp) I then looked up the [Image Method for JUCE](https://docs.juce.com/master/classImage.html) and saw that _Colour | getPixelAt (int x, int y) const_ has a class type of __Colour__, so I then guessed to access getPixelAt() one needed to declare a Colour object and then I looked and saw a method [_toString_](https://docs.juce.com/master/classColour.html)  - From what I understand, in lua you can call methods belonging to certain classes using  __:__ , (_this has to do with a basic type of class structure using tables in lua where table values get passed to functions belonging to the table_)

To be honest I'm not always so lucky as I was with your question. It can be frustrating, but I prefer to see it as an interesting puzzle solving quest! But, to answer your question, the lua to JUCE syntax is still not clear to me.

You can, however, learn a lot from Atom's panels in the DEMO folder of your installation of Ctrlr. There's a lot of information in there, that's for sure. He is obviously on a whole different level.

[__lua__](https://www.rasterbar.com/products/luabind/docs.html) syntax will resemble JUCE (C++) syntax, but it is never the same.

With regard to lua, if you can learn how to use tables, your way of programming in lua will be much much easier.
If you read this over an over (I would print it out) until you know it, your lua skills will increase tenfold! https://www.lua.org/pil/2.5.html

Regards.

**Reply by ghost** [reply]
lua is a different language then C++, some conversions are obvious but some are very complicated. When i doubt i refer to the what() and how() methods in Lua they help a lot. But when writing in Lua always go with the Lua syntax and manual, everything else is just a "wrapper" in Lua

**Reply by bijlevel** [reply]
Thank you again! @dnaldoog you gave me courage, because in essence I'm doing just that: switching all the time between 

- Juce (classes and methods),
- the .def(..) methods of Ctrtlr
- the reference to the LUA language.

 But probably I'm lacking experience in how to make the right connections between these information sources which should result in executable code in Ctrlr. You're right though about the table functionality and exactly in my case I need them a lot!

@rkubiak01 I do use now and then the what()  method to find out about the existence of a class. But when using for instance how() for the Image class, I get a very very long list without, at least for me, seeing any relation to the how.

But I did solve the setPixelAt by looking into your renderer demo panel (thanks for reminding me again John!) and found the code
g:setColour (Colour(0xff0000ff):withAlpha (getOpacity()))

which lead me to the right way of formulating the setPixalAt() function:

frImage:setPixelAt(30, 30, Colour(0xff000000))

and by low and behold at x=30, y=30 coordinates a black pixel appeared in the image!

**Reply by dnaldoog** [reply]
@bijlevel I'm glad you got the results you were looking for. I think you are well on the track to black belt status if you can do all that! Impressed. Perseverance furthers!

**Reply by bijlevel** [reply]
@dnaldoog Thank you!

**Comment by Godlike-Productions** [comment]
what and how are really, really handy.  Just remember to assign a variable and do the what and how on that.  You can't do it on the "command" itself.  

I also search the old forum (or use google and search for "ctrlr something", for example ctrlr wait, to find how to insert a pause into a bulk sysex request loop.

I find these resources pop up a lot (in addition to ctrlr.org forums): https://www.lua.org/manual/5.1/manual.html (ctrlr uses 5.1.  Some of the bitwise functions are done differently. 5.2 uses >> for example, 5.1 uses a library), http://lua-users.org/wiki/LuaTypeChecking lots of really useful examples there.

Oh and previous panels.

I'm going to start puting code snippets up here as well, such as the class definitions that I'm using for the Jupiter 8 Librarian, and other problems that I've solved.

**Reply by bijlevel** [reply]
Thank you for these tips and pointing out valuable resources for LUA!

**Comment by dobo365** [comment]
Yep, it takes time to go deeper into Ctrlr with more complex things and indeed you listed the 3 sources of info:
- the Ctrlr forum
- Juce API site 
- Lua programming sites as lua.org or others
- Demo panels + other people panels
- Ctrlr code
Usually this last one is not needed but indeed it is good to have what() and how()
The thing is also to once in a while dig into a new subject so you keep finding new possibilities.
Challenge yourself then take the time to read many things and to be patient with bugs in your code ;-)

**Comment by Godlike-Productions** [comment]
There is some follow up, and an alternative method using filled rectangles over at the old forums.

https://ctrlr.org/forums/topic/creating-images-using-lua/

Hopefully I'll have my Jupiter 8 panel finished soon and can start working on this with you.

**Reply by bijlevel** [reply]
Hi Jason,
I believe you already did send (me) this link. In the meantime I managed to display the graphic screen of the K2500 in the panel and I think the example of Dnaldoog (LCD-xpm-Matrix) in the thread is the key to displaying the text part as pixels. At the moment I'm studying it. I think it is then also possible to display the pixeltext inverted when needed, exactly the way Kurzweil does it. I'm happy you're still offering to help. Should we keep this forum informed about our progress, or keep it private? Good luck with your Jupiter! (It's now bedtime in my country...)

**Reply by Godlike-Productions** [reply]
On the Kurzweil's this is handled with the font (it has normal and inverted). I have a copy of the font file somewhere, but it's very similar to the  ZX81 font included in ctrlr, which also has inverted fonts. I think we should just do text on a seperate layer, the same way the Kurzweil's do. We could do a different colour the same way it was done on the java editor

**Reply by bijlevel** [reply]
The ZX81 font does indeed look like the one used on the Kurzweil. But (again!) I did run into bad behaviour of Ctrlr. I did assign this font to an existing label, but it refuses to change to that font, even when reloading the panel. BTW when assigning the font to a new label, the default is "inverted" and I don't see where I can change it to normal...

The real problem, as I see it, is not displaying the text on a separate layer of whatever, but to find out where exactly to place the text of the active patch or in general a parameter (and its value) in relation to the graphics layer. It is possible to find out which parameter and value is active. Kurzweil does provide the sysex for that info, but ATM I don't see a way how to place that text (inverted) above the graphics layer, in a way it coincidences with that layer. But I keep on studying this "problem". As I wrote I did manage to display the graphics layer in the panel, but it doesn't look yet exactly the same as on the Kurzweil LCD. I don't know why, have also to find out. Enough to do for the weekend and maybe you have some ideas!
